### PR TITLE
SNOW-1818230: add concat_ws_ignore_nulls

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1818230

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   - Add `concat_ws_ignore_nulls` that ignores nulls and runs `concat_ws` operation.
   - Update `substring` function to make len argument optional to align with snowflake functionality.

### CODE REVIEW

File: src/snowflake/snowpark/functions.py [hunk 1]
```diff
@@ -2628,18 +2628,28 @@ def substring(
...
+        ... )
+        >>> df.select(substring(col("S"), 2)).collect()
+        [Row(SUBSTRING("S", 2)='bc'), Row(SUBSTRING("S", 2)='ef')]
     """
     s = _to_col_if_str(str, "substring")
     p = pos if isinstance(pos, Column) else lit(pos)
-    length = len if isinstance(len, Column) else lit(len)
-    return builtin("substring")(s, p, length)
+    if len is not None:
+        length = len if isinstance(len, Column) else lit(len)
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/functions.py:2650]: I've seen other places in function do this with `Column._to_expr(len)`.

File: src/snowflake/snowpark/functions.py [hunk 2]
```diff
@@ -2935,6 +2945,39 @@ def concat_ws(*cols: ColumnOrName) -> Column:
...
+        |Hello,World                                       |
+        |                                                  |
+        |Hello                                             |
+        ----------------------------------------------------
+        <BLANKLINE>
+    """
+    columns = [_to_col_if_str(c, "concat_ws_ignore_nulls") for c in cols]
+    names = ",".join([c.get_name() for c in columns])
+    input_column_array = array_construct_compact(*columns)
+    reduced_result = builtin("reduce")(
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/functions.py:2973]: reduce is to return a single value. Should it be filter?

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/functions.py:2973]: Another way is to use `concat_was(filter(...))` to remove the nulls first then concat.

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/functions.py:2972]: Return value of `filter` is an array. `concat_ws` does not take `ARRAY` as an input therefore we had to use reduce.

**Review by sfc-gh-jrose — APPROVED** [review]

File: src/snowflake/snowpark/functions.py [hunk 3]
```diff
@@ -3030,18 +3030,27 @@ def substring(
...
+        >>> df.select(substring(col("S"), 2)).collect()
+        [Row(SUBSTRING("S", 2)='bc'), Row(SUBSTRING("S", 2)='ef')]
     """
     s = _to_col_if_str(str, "substring")
     p = pos if isinstance(pos, Column) else lit(pos, _emit_ast=_emit_ast)
-    length = len if isinstance(len, Column) else lit(len, _emit_ast=_emit_ast)
-    return builtin("substring", _emit_ast=_emit_ast)(s, p, length)
+    if len is None:
+        return builtin("substring", _emit_ast=_emit_ast)(s, p)
+    return builtin("substring", _emit_ast=_emit_ast)(s, p, Column._to_expr(len))
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:3053]: The `Column._to_expr` may cause issues when we add the AST here. We use instead the following and make it explicit:

```
len if isinstance(len, Column) else lit(len)
```

This should also have the correct type then, i.e. not using Expression objects to be passed to the callable created by builtin. Can we change to that?

Also, we're happy to help with adding the AST here. Can we file a ticket onto https://snowflakecomputing.atlassian.net/browse/SNOW-1489960 and add TODO in code with the ticket number to track the work?

If you'd like to give the AST directly a shot, happy to help out and walk you through the process - else also fine with JIRA ticket + TODO and our team will take care of it. Whichever works for you!

File: src/snowflake/snowpark/functions.py [hunk 4]
```diff
@@ -3392,6 +3401,42 @@ def concat_ws(*cols: ColumnOrName, _emit_ast: bool = True) -> Column:
...
+        ----------------------------------------------------
+        |Hello,World                                       |
+        |                                                  |
+        |Hello                                             |
+        ----------------------------------------------------
+        <BLANKLINE>
+    """
+    columns = [_to_col_if_str(c, "concat_ws_ignore_nulls") for c in cols]
+    names = ",".join([c.get_name() for c in columns])
+    input_column_array = array_construct_compact(*columns, _emit_ast=_emit_ast)
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:3431]: you can do `_emit_ast=False` here. We'll capture the AST top-level in the body of this function. Basically the thumb rule should be whenever calling some other public function as part of the logic, use `_emit_ast=False`.

File: src/snowflake/snowpark/functions.py [hunk 5]
```diff
@@ -3392,6 +3401,42 @@ def concat_ws(*cols: ColumnOrName, _emit_ast: bool = True) -> Column:
...
+    reduced_result = builtin("reduce", _emit_ast=_emit_ast)(
+        input_column_array, lit(""), sql_expr(f"(l, r) -> l || '{sep}' || r")
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:3433]: same here, prob. _emit_ast=False is correct.

File: src/snowflake/snowpark/functions.py [hunk 6]
```diff
@@ -3392,6 +3401,42 @@ def concat_ws(*cols: ColumnOrName, _emit_ast: bool = True) -> Column:
...
+    )
+    return substring(reduced_result, 2, _emit_ast=_emit_ast).alias(
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:3435]: you can use _emit_ast=False for both, we'll need to explicitly create the AST for this and then assign to the columnar expression returned.

**Review by sfc-gh-lspiegelberg — COMMENTED** [review]
left AST related comments

File: src/snowflake/snowpark/functions.py (hunk 3 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/functions.py:3053]: This is the same usage pattern as `to_date`. Shouldn't that also create problems?

Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/functions.py:3053]: We fixed a lot of the `Column._to_expr(...)` cases. There may be some that fell through, merging long-standing feature branches is hard. But yes, we should not use `Column._to_expr(...)` anywhere. Better to be explicit. Happy to discuss/share more in detail.

File: src/snowflake/snowpark/functions.py (hunk 1 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/functions.py]: we should not use `Column._to_expr(len)`. Happy to discuss why, it also clashes with the current AST design.

File: src/snowflake/snowpark/functions.py [hunk 7]
```diff
@@ -3030,16 +3030,26 @@ def substring(
...
+
+    Example 2::
+        >>> df = session.create_dataframe(
+        ...     ["abc", "def"],
+        ...     schema=["S"],
+        ... )
+        >>> df.select(substring(col("S"), 2)).collect()
+        [Row(SUBSTRING("S", 2)='bc'), Row(SUBSTRING("S", 2)='ef')]
     """
     s = _to_col_if_str(str, "substring")
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:3049]: can we add `# TODO: SNOW-1831917 create ast`?

**Review by sfc-gh-lspiegelberg — APPROVED** [review]
LGTM (maybe add todo also for substring API, it likely will already work but not sure we have a test for the expanded version here)

File: src/snowflake/snowpark/functions.py (hunk 7 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/functions.py:3049]: this is not a new API. I am only updating the signature here

**Pull request merged**